package.json

```diff
@@ -23,8 +23,12 @@
     "typescript": "3.2.4"
   },
   "devDependencies": {
+    "@babel/core": "7.2.2",
+    "@storybook/react": "4.1.11",
     "@types/enzyme": "^3.1.15",
     "@types/enzyme-adapter-react-16": "^1.0.3",
+    "@types/storybook__react": "4.0.0",
+    "babel-loader": "8.0.4",
```

Comment:
I am worried that this could get out of sync with [create-react-app](https://github.com/facebook/create-react-app/blob/714599dae652426331007f95cee5e26ea627a1b1/packages/react-scripts/package.json#L32). I'm not totally sure if that's a problem or not.

Reply:
This was a bit tricky to deal with. If I don’t include `babel-loader` in `package.json` then Storybook gives the following error:

```
yarn storybook
yarn run v1.13.0
$ start-storybook -p 9001 -c stories/setup
info @storybook/react v4.1.11
info
info => Loading presets
info => Loading presets
ERR! Error: Cannot find module 'babel-loader/package.json' from '/Users/bsilverberg/Documents/gitRepos/addons-code-manager'
ERR!     at module.exports (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/resolve/lib/sync.js:58:15)
ERR!     at isBabelLoader8 (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/utils/load-custom-babel-config.js:68:52)
ERR!     at _default (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/utils/load-custom-babel-config.js:88:10)
ERR!     at Object.babel (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/common/common-preset.js:19:45)
ERR!     at accumulationPromise.then.newConfig (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/presets.js:78:72)
ERR!     at <anonymous>
ERR!     at process._tickCallback (internal/process/next_tick.js:188:7)
ERR!  { Error: Cannot find module 'babel-loader/package.json' from '/Users/bsilverberg/Documents/gitRepos/addons-code-manager'
ERR!     at module.exports (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/resolve/lib/sync.js:58:15)
ERR!     at isBabelLoader8 (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/utils/load-custom-babel-config.js:68:52)
ERR!     at _default (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/utils/load-custom-babel-config.js:88:10)
ERR!     at Object.babel (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/common/common-preset.js:19:45)
ERR!     at accumulationPromise.then.newConfig (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/presets.js:78:72)
ERR!     at <anonymous>
ERR!     at process._tickCallback (internal/process/next_tick.js:188:7)
ERR!   stack: 'Error: Cannot find module \'babel-loader/package.json\' from \'/Users/bsilverberg/Documents/gitRepos/addons-code-manager\'\n    at module.exports (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/resolve/lib/sync.js:58:15)\n    at isBabelLoader8 (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/utils/load-custom-babel-config.js:68:52)\n    at _default (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/utils/load-custom-babel-config.js:88:10)\n    at Object.babel (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/common/common-preset.js:19:45)\n    at accumulationPromise.then.newConfig (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/presets.js:78:72)\n    at <anonymous>\n    at process._tickCallback (internal/process/next_tick.js:188:7)',
ERR!   code: 'MODULE_NOT_FOUND' }
(node:97725) UnhandledPromiseRejectionWarning: Error: Cannot find module 'babel-loader/package.json' from '/Users/bsilverberg/Documents/gitRepos/addons-code-manager'
    at module.exports (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/resolve/lib/sync.js:58:15)
    at isBabelLoader8 (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/utils/load-custom-babel-config.js:68:52)
    at _default (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/utils/load-custom-babel-config.js:88:10)
    at Object.babel (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/common/common-preset.js:19:45)
    at accumulationPromise.then.newConfig (/Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/@storybook/core/dist/server/presets.js:78:72)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
(node:97725) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 5)
(node:97725) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

If I add the default version of `babel-loader`, which is version `8.0.5`, then Storybook is happy, but when I run `yarn test` or `yarn dev`, I see the following:

```
yarn dev
yarn run v1.13.0
$ cat .env.common-local .env.dev > .env.local && yarn start-dev
$ concurrently --kill-others 'PORT=3100 BROWSER=none react-scripts start' 'NODE_ENV=development yarn serve'
[0]
[0] There might be a problem with the project dependency tree.
[0] It is likely not a bug in Create React App, but something you need to fix locally.
[0]
[0] The react-scripts package provided by Create React App requires a dependency:
[0]
[0]   "babel-loader": "8.0.4"
[0]
[0] Don't try to install it manually: your package manager does it automatically.
[0] However, a different version of babel-loader was detected higher up in the tree:
[0]
[0]   /Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/babel-loader (version: 8.0.5)
[0]
[0] Manually installing incompatible versions is known to cause hard-to-debug issues.
[0]
[0] If you would prefer to ignore this check, add SKIP_PREFLIGHT_CHECK=true to an .env file in your project.
[0] That will permanently disable this message but you might encounter other issues.
[0]
[0] To fix the dependency tree, try following the steps below in the exact order:
[0]
[0]   1. Delete package-lock.json (not package.json!) and/or yarn.lock in your project folder.
[0]   2. Delete node_modules in your project folder.
[0]   3. Remove "babel-loader" from dependencies and/or devDependencies in the package.json file in your project folder.
[0]   4. Run npm install or yarn, depending on the package manager you use.
[0]
[0] In most cases, this should be enough to fix the problem.
[0] If this has not helped, there are a few other things you can try:
[0]
[0]   5. If you used npm, install yarn (http://yarnpkg.com/) and repeat the above steps with it instead.
[0]      This may help because npm has known issues with package hoisting which may get resolved in future versions.
[0]
[0]   6. Check if /Users/bsilverberg/Documents/gitRepos/addons-code-manager/node_modules/babel-loader is outside your project directory.
[0]      For example, you might have accidentally installed something in your home folder.
[0]
[0]   7. Try running npm ls babel-loader in your project folder.
[0]      This will tell you which other package (apart from the expected react-scripts) installed babel-loader.
[0]
[0] If nothing else helps, add SKIP_PREFLIGHT_CHECK=true to an .env file in your project.
[0] That would permanently disable this preflight check in case you want to proceed anyway.
[0]
[0] P.S. We know this message is long but please read the steps above :-) We hope you find them helpful!
[0]
[0] PORT=3100 BROWSER=none react-scripts start exited with code 1
--> Sending SIGTERM to other processes..
[1] NODE_ENV=development yarn serve exited with code 1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Running `npm ls babel-loader`, without any reference to `babel-loader` in `package.json` yields this:

```
npm ls babel-loader
addons-code-manager@0.1.0 /Users/bsilverberg/Documents/gitRepos/addons-code-manager
├─┬ @storybook/react@4.1.11
│ └─┬ babel-preset-react-app@6.1.0
│   └── babel-loader@8.0.4
└─┬ react-scripts@2.1.3
  ├── babel-loader@8.0.4
  └─┬ babel-preset-react-app@7.0.0
    └── babel-loader@8.0.4  deduped
```

Based on that, I tried adding version `8.0.4` of `babel-loader` to `package.json` and then everything worked.

I’m not sure why Storybook is complaining about a missing `babel-loader` when one clearly exists in `node_modules`, but it seems like it wants it to specifically be in `package.json` (based on the error message). I also found this: https://github.com/storybooks/storybook/issues/5183#issuecomment-454333915 which suggests doing the same thing as a workaround.

Reply:
> I also found this: storybooks/storybook#5183 (comment) which suggests doing the same thing as a workaround.

Oh, many, that's unfortunate. Well, it looks like the patch is ready to land upstream so hopefully we can remove this soon.